### ISSUE TITLE
Integrating NSSM into a CasADi Optimization

### ISSUE THREAD

**Issue by lclemen** [pnnl/neuromancer#228]
I am trying to integrate the output from a trained NSSM into a CasADi optimization as a dynamic constraint of the type

T_{k+1} == NSSM(T_k, u_k, d_k)

I have run into issues with both the low level (nlpsol) and high level Opti interfaces with this. 

When using the Opti interface, unfortunately this requires a gradient output and changing to a differential based solver like the SQPsolver did not work and I got an error.

When using the nlpsol approach I have run into similar issues as well as importing and building a CasADi function from the trained NSSM model.

Here is an example from the nlpsol approach:

```
# --- NSSMWrapper Class WITH ADJOINT ---
class NSSMWrapper(Callback):
    def __init__(self, name='nssm_wrapper', nx=None, nu=None, nd=None, eval_func=None):
        super().__init__()
        self.nx = nx
        self.nu = nu
        self.nd = nd
        self.eval_func = eval_func # PyTorch model/function
        self.construct(name)
        # Keep track of tensors from last eval for gradient calculation
        self.last_t = None
        self.last_u = None
        self.last_d = None
        self.last_output = None

    def get_n_in(self): return 3
    def get_n_out(self): return 1

    def get_sparsity_in(self, idx):
        if idx == 0: return Sparsity.dense(1, self.nx) # t
        elif idx == 1: return Sparsity.dense(1, self.nu) # u
        elif idx == 2: return Sparsity.dense(self.nd, 1) # d
        else: raise IndexError("Input index out of bounds")

    def get_sparsity_out(self, idx):
        if idx == 0: return Sparsity.dense(self.nx, 1) # temp_next
        else: raise IndexError("Output index out of bounds")

    # Implement eval and store tensors for gradient calculation
    def eval(self, args):
        t_dm = args[0]
        u_dm = args[1]
        d_dm = args[2]

        # Convert DM to PyTorch tensors, ENSURE requires_grad=True for inputs
        # that derivatives are needed for (t and u).
        self.last_t = torch.tensor(np.array(t_dm), requires_grad=True, dtype=torch.float32).reshape(1, self.nx)
        self.last_u = torch.tensor(np.array(u_dm), requires_grad=True, dtype=torch.float32).reshape(1, self.nu)
        # Disturbance usually doesn't require grad from the optimizer's perspective
        self.last_d = torch.tensor(np.array(d_dm), requires_grad=False, dtype=torch.float32).reshape(self.nd, 1)

        # Use the separate method for the actual PyTorch call
        output_np = self.dynamic_constraint_torch(self.last_t, self.last_u, self.last_d)
        # Store the output tensor as well (needed for autograd.grad)
        # Ensure output_np is correctly converted back to tensor if dynamic_constraint_torch returns numpy
        if isinstance(output_np, np.ndarray):
             self.last_output = torch.tensor(output_np, dtype=torch.float32)
        else: # Assume it's already a tensor
             self.last_output = output_np

        # Return numerical result as DM
        return [DM(self.last_output.detach().numpy().reshape(self.nx, 1))]

    # Separated torch logic for clarity
    def dynamic_constraint_torch(self, t_tensor, u_tensor, d_tensor):
        # t_tensor shape (1, nx), u_tensor shape (1, nu), d_tensor shape (nd, 1)
        # Reshape d if model expects (1, nd)
        d_tensor_reshaped = d_tensor.T # (1, nd)
        with torch.no_grad(): # Still no_grad for normal evaluation pass
            inputs = {
                # Add batch dim if model expects it explicitly
                "X": t_tensor.unsqueeze(0),
                "U": u_tensor.unsqueeze(0),
                "D": d_tensor_reshaped.unsqueeze(0)
            }
            outputs = self.eval_func(inputs)
            # Return the raw output tensor before detaching for gradient calculation later
            return outputs["Y_pred"].squeeze(0) # Remove batch dim, shape (nx,)

    # --- Derivative Implementation ---
    def has_adjoint(self):
        # Indicate that we provide adjoint derivatives
        return True

    def get_adjoint(self, name, adj_i, adj_o, opts):
        # adj_o[0] is the sensitivity of the output (dL/d(temp_next)) from CasADi (DM)
        # We need to compute dL/dt and dL/du using VJP

        # 1. Convert CasADi adjoint seed to PyTorch tensor
        # Ensure shape matches the output tensor self.last_output (likely (nx,))
        adj_seed_tensor = torch.tensor(np.array(adj_o[0]).flatten(), dtype=torch.float32)

        # 2. Ensure the tensors from the last eval call still exist and require grad
        if self.last_t is None or self.last_u is None or self.last_output is None:
            # This shouldn't happen if eval was called right before get_adjoint
            raise RuntimeError("Inputs/Outputs from previous eval not stored for adjoint calculation.")
        if not self.last_t.requires_grad or not self.last_u.requires_grad:
             raise RuntimeError("Input tensors do not require grad for adjoint calculation.")


        # 3. Compute VJPs using torch.autograd.grad
        # Need grad of the output wrt inputs t and u
        grads = torch.autograd.grad(
            outputs=self.last_output,       # Output tensor from forward pass
            inputs=(self.last_t, self.last_u), # Inputs requiring gradient
            grad_outputs=adj_seed_tensor,   # The seed vector (dL/dOutput)
            retain_graph=False,             # Usually False unless needed elsewhere
            allow_unused=True               # Important if an input doesn't affect output
        )

        adj_t = grads[0] # dL/dt = (dL/dOutput) * (dOutput/dt)
        adj_u = grads[1] # dL/du = (dL/dOutput) * (dOutput/du)

        # Handle cases where gradient might be None (if input didn't affect output)
        adj_t_dm = DM(adj_t.numpy().reshape(1, self.nx)) if adj_t is not None else DM.zeros(1, self.nx)
        adj_u_dm = DM(adj_u.numpy().reshape(1, self.nu)) if adj_u is not None else DM.zeros(1, self.nu)

        # Return adjoints for t, u, and d (d's is zero as it's a parameter here)
        # Shapes must match get_sparsity_in
        return [adj_t_dm, adj_u_dm, DM.zeros(self.nd, 1)]


self.thermal_model.dynamics_model.eval()
    nssm_callback: Callback = NSSMWrapper(nx=nx, nu=nu, nd=nd,
                                           eval_func=self.thermal_model.dynamics_model,
                                           name='nssm_cb') # Give a name

    # --- Define Symbolic Placeholders ---
    sym_t = MX.sym('t_in', 1, nx)
    sym_u = MX.sym('u_in', 1, nu)
    sym_d = MX.sym('d_in', nd, 1)

    # --- Symbolic call using callback's .call() method ---
    symbolic_output_list = nssm_callback.call([sym_t, sym_u, sym_d])
    symbolic_output_expr = symbolic_output_list[0]

    # --- Create the CasADi Function (NO FD options needed now) ---
    nssm_func = Function(
        'nssm_func',
        [sym_t, sym_u, sym_d],   # Inputs
        [symbolic_output_expr]   # Outputs
        # No fd_opts needed, relies on has_adjoint/get_adjoint now
    )
    print("CasADi function created successfully (using callback adjoint).")
    print("Function signature:", nssm_func)


# Dynamics and Power Constraints
    for step in range(horizon):
        t_k = temperature[step, :]         # Symbolic MX (1, nx)
        u_k = inputs[step, :]            # Symbolic MX (1, nu)
        d_k_sym = dist_param_matrix[:, step] # Symbolic MX (nd, 1) slice of p

        # Call the function using positional list argument
        outputs_list = nssm_func([t_k, u_k, d_k_sym]) # Uses get_adjoint internally now
        temperature_next = outputs_list[0]

        # Dynamics constraint: T_k+1 - prediction = 0
        g += [temperature[step + 1, :].T - temperature_next]

        # Power balance constraint
        hvac_power_start_idx = 4 * nx
        power_columns_k = inputs[step, hvac_power_start_idx:]
        hvac_total_power = sum1(reshape(power_columns_k, nx, 1))
        baseline = self.optimization_data.processed_grid_data["baseline_power_kw"].values[step]
        solar = self.optimization_data.processed_grid_data["solar_power_kw"].values[step]
        ev = self.optimization_data.processed_grid_data["ev_charger_power_kw"].values[step]
        g += [interval_power[step] - (hvac_total_power + baseline + solar + ev)]

    g = vertcat(*g)
    print(f"Total constraints: {g.shape}")

    # Constraint bounds
    num_g_constraints = horizon * (nx + 1)
    lbg = [0] * num_g_constraints
    ubg = [0] * num_g_constraints 
```
It seems like it might be best to build the NSSM again in CasADi functions and then use that. Is that the recommended path forward?

**Comment by drgona** [comment]
Hi @lclemen, even though this is an interesting case, this is not really a Neuromancer-related issue.
Neuromancer does not support model export to CasADi.

They are different ways to approach this, one is to build exactly the same computational graph in CasADi and load the parameters from the trained model in Neuromancer/Pytorch.

The other is to use l4casadi library that supports linearization of Pytorch models that can be used as models in CasADi.
https://github.com/Tim-Salzmann/l4casadi
